rfcs/20230622-quantized-reduction.md

```diff
@@ -0,0 +1,372 @@
+like add, this means that the the result of accumulation can overflow in which
+case the result will be implementation defined (e.g.,
+        [saturated](https://en.wikipedia.org/wiki/Saturation_arithmetic) or
+        [wrap around](https://en.wikipedia.org/wiki/Integer_overflow)).  From
+the conversation with customers it seems a reasonable behavior for non quantized
+data types. However, with quantized data types, such loss in precision is not
+acceptable and hence the motivation is to perform the accumulation in some
+higher data type.
+
+The RFC introduces the following proposal, emerged out of discussion in the
+[thread](https://github.com/openxla/stablehlo/pull/1538#issuecomment-1599476906)
+, along with their tradeoffs.
+
+The proposal allows the reducer block to express the computation in a different
+element type (preferably wider accumulation type) than the one used in reduce
+op's ops arguments and return type. For illustrative purposes, in the following
+example, the operand element type
+`tensor<!quant.uniform<ui8:f32, input_scale:input_zp>>`  is different from the
+element type for reduction region's block arguments.  Similarly, the element
+type of the reduce op's result
+`!quant.uniform<ui8:f32, output_scale:output_zp>>` is different from that of
+block return (`tensor<!quant.uniform<i32:f32, accum_scale:accum_zp>>`).
+
+```mlir
+%result = "stablehlo.reduce"(%input, %init_value) ({
+    ^reduce_computation(
+            %lhs: tensor<!quant.uniform<i32:f32, accum_scale:accum_zp>>,
+            %rhs: tensor<!quant.uniform<i32:f32, accum_scale:accum_zp>>):
+        %add = "stablehlo.add"(%lhs, %rhs)
+            : (tensor<!quant.uniform<i32:f32, accum_scale:accum_zp>>,
+               tensor<!quant.uniform<i32:f32, accum_scale:accum_zp>>)
+            -> tensor<!quant.uniform<i32:f32, accum_scale:accum_zp>>
+        "stablehlo.return"(%add)
+            : (tensor<!quant.uniform<i32:f32, accum_scale:accum_zp>>) -> ()
+    }) {
+        dimensions = dense<1> : tensor<i64>
+    } : (tensor<5 x 1 x !quant.uniform<ui8:f32, input_scale:input_zp>>,
+         tensor<!quant.uniform<ui8:f32, input_scale:input_zp>>)
+    -> tensor<5 x !quant.uniform<ui8:f32, output_scale:output_zp>>
+```
+
+### Semantics
+
+If (1) the input operand type is different from the reduction block
+argument type or (2) the op result type is different from the reduction block
+return type, there will be implicit type conversion defined by either
+`stablehlo.convert`, `stablehlo.uniform_quantize`, or
+`stablehlo.uniform_dequantize`. For example,
+
+ | Implicit type conversion op       | element type of operand or result type | element type of block argument or block return type |
+ |-----------------------------------|----------------------------------------|-----------------------------------------------------|
+ | (A) `stablehlo.uniorm_quantize`   | quantized tensor                       | quantized tensor                                    |
```

Comment:
Nit: typo `uniform_quantize'. Also, of the result type is quantized but the block is float, the conversion would be `uniform_quantize'. The table seems to specify argument conversions.